pkg/edgehub/controller_test.go

```
@@ -0,0 +1,820 @@
+/*
+Copyright 2019 The KubeEdge Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package edgehub
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"reflect"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	"github.com/pkg/errors"
+
+	"github.com/kubeedge/kubeedge/beehive/pkg/core"
+	"github.com/kubeedge/kubeedge/beehive/pkg/core/context"
+	"github.com/kubeedge/kubeedge/beehive/pkg/core/model"
+	"github.com/kubeedge/kubeedge/mocks/edgehub"
+	"github.com/kubeedge/kubeedge/pkg/common/util"
+	_ "github.com/kubeedge/kubeedge/pkg/devicetwin"
+	"github.com/kubeedge/kubeedge/pkg/edgehub/config"
+)
+
+const (
+	CertFile = "/tmp/kubeedge/Certs/edge.crt"
+	KeyFile  = "/tmp/kubeedge/Certs/edge.key"
+)
+
+//TestServer is a fake http server created for testing
+var TestServer *httptest.Server
+
+// mockAdapter is a mocked adapter implementation
+var mockAdapter *edgehub.MockAdapter
+
+//init() starts the test server and generates test certificates for testing
+func init() {
+	newTestServer()
+	err := util.GenerateTestCertificate("/tmp/kubeedge/Certs/", "edge", "edge")
+	if err != nil {
+		fmt.Println("Error in creating fake certificates: ", err)
```

panic instead of print